### PR TITLE
feat(config): Support env variables and command line args

### PR DESCRIPTION
**Pull request by RaduW**
WIP --

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -40,9 +40,11 @@ relay-general = { path = "relay-general" }
...
 serde = "1.0.98"
 serde_json = "1.0.40"
+serde_yaml = "0.8.9"
 
 [target."cfg(not(windows))".dependencies]
 openssl-probe = "0.1.2"
 
 [build-dependencies]
 clap = { version = "2.33.0", default-features = false }
+insta = { version = "0.12.0", features =  ["ron", "redactions"] }
```
Comment (jan-auer) [inline Cargo.toml]: Can you back out the changes to `/Cargo.toml`? `serde_yaml` is already a dependency of `relay-config`.

If you need `insta`, please use version `0.15.0` which is used in other crates, and move the dependency to `relay-config/Cargo.toml`.

File: relay-config/src/config.rs [hunk 2]
```diff
@@ -634,6 +669,47 @@ impl Config {
...
+        }
+
+        let processing = &mut self.values.processing;
+        if let Some(enabled) = overrides.processing {
+            processing.enabled = enabled;
+        }
+
+        //TODO finish here
+        if let Some(redis) = overrides.redis_url {
+            //processing.redis = from_str(redis.as_str());
```
Comment (jan-auer) [inline relay-config/src/config.rs]: Please just support a single node configuration via the env var. Also, please remove the `serde_yaml` dependency. Same for Kafka, we just support a single bootstrap node with no special configuration.

```rust
processing.redis = RedisConfig::Single(redis);
```

Then, the env var can be set to:

```
export RELAY_REDIS_URL="redis://127.0.0.1:4711"
```

File: src/cli.rs [hunk 3]
```diff
@@ -338,7 +338,7 @@ pub fn process_event<'a>(matches: &ArgMatches<'a>) -> Result<(), Error> {
     Ok(())
 }
 
-pub fn run<'a>(config: Config, _matches: &ArgMatches<'a>) -> Result<(), Error> {
+pub fn run<'a>(mut config: Config, _matches: &ArgMatches<'a>) -> Result<(), Error> {
```
Comment (jan-auer) [inline src/cli.rs]: Does this need to be mut?

File: src/main.rs [hunk 4]
```diff
@@ -1,5 +1,6 @@
 mod cli;
 mod cliapp;
+mod env_arg_override;
```
Comment (jan-auer) [inline src/main.rs]: Can we use a shorter name, like `env` or `env_args`?

**Comment by jan-auer** [comment]
CI failures are this:
```
error: Invalid override (field incomplete credentials:id=None public_key=None secret_keyNone)
```

File: relay-config/src/config.rs [hunk 5]
```diff
@@ -643,6 +700,160 @@ impl Config {
         Ok(config)
     }
 
+    /// Override configuration with values coming from other sources (e.g. env variables or
+    /// command line parameters)
+    pub fn override_config(
```
Comment (jan-auer) [inline relay-config/src/config.rs]: _nit_: This function is defined on `Config`, so the `_config` suffix is redundant.

```suggestion
    pub fn override(
```

File: relay-config/src/config.rs [hunk 6]
```diff
@@ -11,39 +11,61 @@ use failure::{Backtrace, Context, Fail};
...
-    pub fn filename(&self) -> &Path {
-        &self.filename
+    fn new_field_error<NameT, ValueT>(
+        name: NameT,
+        value: ValueT,
+        inner: Context<ConfigErrorKind>,
+    ) -> Self
+    where
+        NameT: Into<String>,
+        ValueT: Into<String>,
```
Comment (jan-auer) [inline relay-config/src/config.rs]: It seems that the field name is (and should) always be a `&'static str`. Values are always effectively strings. Can we simplify this signature then?

I'm wondering if we actually need the original value. The inner error will usually contain sufficient information to see what the problem is. Especially for values like the public/private keys, it may not make sense to print them out again. Personally, I would just retain the field name an inner error.

Finally, since this type is an error, so the function name can be simplified.

```rust
fn for_field(name: &'static str, value: String, ...)
```

File: relay-config/src/config.rs [hunk 7]
```diff
@@ -87,6 +116,34 @@ pub enum ConfigErrorKind {
...
+    /// parsed into the proper type
+    #[fail(display = "Invalid override")]
+    InvalidOverride,
+}
+
+/// Structure used to hold information about configuration overrides via
+/// CLI parameters or environment variables
+#[derive(Debug, Default)]
+pub struct OverridableConfig {
+    /// the upstream for the Relay
```
Comment (jan-auer) [inline relay-config/src/config.rs]: _nit_: Can you use the same doc strings here as on the original fields?

File: relay-config/src/config.rs [hunk 8]
```diff
@@ -87,6 +116,34 @@ pub enum ConfigErrorKind {
...
+    pub upstream: Option<String>,
+    /// the host
+    pub host: Option<String>,
+    /// the port serving requests
+    pub port: Option<String>,
+    /// "true" if processing is enabled "false" otherwise
+    pub processing: Option<String>,
```
Comment (jan-auer) [inline relay-config/src/config.rs:121]: Would it be possible to use an actual `bool` here?

File: relay-config/src/config.rs [hunk 9]
```diff
@@ -643,6 +700,160 @@ impl Config {
...
+        overrides: OverridableConfig,
+    ) -> Result<&mut Self, ConfigError> {
+        let relay = &mut self.values.relay;
+
+        if let Some(upstream) = overrides.upstream {
+            relay.upstream = UpstreamDescriptor::from_str(&upstream).map_err(|err| {
+                ConfigError::new_field_error(
+                    "upstream",
+                    upstream,
+                    err.context(ConfigErrorKind::InvalidOverride),
```
Comment (jan-auer) [inline relay-config/src/config.rs]: The calls to `context` seem redundant. You could pull them into `ConfigError::new_field_error`. The signature would be generic over `E: Display + Send + Sync + 'static`.

File: relay-config/src/config.rs [hunk 10]
```diff
@@ -643,6 +700,160 @@ impl Config {
...
+                (None, None, None) => {
+                    // nothing provided, we'll just leave the credentials None, maybe we
+                    // don't need them in the current command or we'll override them latter
+                }
+                (id, public_key, secret_key) => {
+                    return Err(ConfigError::new_field_error(
+                        "incomplete credentials",
+                        format!(
+                            "id={:?} public_key={:?} secret_key={:?}",
+                            id, public_key, secret_key
```
Comment (jan-auer) [inline relay-config/src/config.rs]: I think this is pretty excessive. It would be nicer to just print out which field (e.g. relay id, public key, ...) is missing, but not the value. Actually, in all cases.

File: src/cli.rs [hunk 11]
```diff
@@ -50,19 +50,63 @@ pub fn execute() -> Result<(), Error> {
...
+    }
+}
+
+/// Extract config arguments from environment variables
+pub fn extract_config_env_vars() -> OverridableConfig {
+    OverridableConfig {
+        upstream: env::var("RELAY_UPSTREAM_URL").ok(),
+        host: env::var("RELAY_HOST").ok(),
+        port: env::var("RELAY_PORT").ok(),
+        processing: env::var("RELAY_PROCESSING_ENABLED").ok(),
```
Comment (jan-auer) [inline src/cli.rs:101]: Here, it might also be desirable to allow `RELAY_PROCESSING_ENABLED=1`. Anything other than `"true"` or `"1"` should be considered `false`, even an empty string.

File: src/cliapp.rs [hunk 12]
```diff
@@ -26,10 +26,94 @@ pub fn make_app() -> App<'static, 'static> {
...
+                )
+                .arg(
+                    Arg::with_name("id")
+                        .long("id")
+                        .short("i")
+                        .value_name("RELAY_ID")
+                        .help("The relay ID to set"),
+                )
+                .arg(
+                    Arg::with_name("upstream-url")
```
Comment (jan-auer) [inline src/cliapp.rs]: Can we shorten this to `--upstream`?

File: src/cliapp.rs [hunk 13]
```diff
@@ -26,10 +26,94 @@ pub fn make_app() -> App<'static, 'static> {
...
+                .arg(
+                    Arg::with_name("port")
+                        .value_name("PORT")
+                        .takes_value(true)
+                        .short("P")
+                        .long("port")
+                        .help("The server port."),
+                )
+                .arg(
+                    Arg::with_name("processing-enabled")
```
Comment (jan-auer) [inline src/cliapp.rs]: Can we shorten this to `--processing`?

File: src/cliapp.rs [hunk 14]
```diff
@@ -26,10 +26,94 @@ pub fn make_app() -> App<'static, 'static> {
...
+                        .short("e")
+                        .long("processing-enabled")
+                        .help("Enable processing."),
+                )
+                .arg(
+                    Arg::with_name("processing-disabled")
```
Comment (jan-auer) [inline src/cliapp.rs]: Since this is the default, I'd remove this. 

If you strongly prefer to keep it, please add `.conflicts_with("processing_enabled")`

File: src/cliapp.rs [hunk 15]
```diff
@@ -26,10 +26,94 @@ pub fn make_app() -> App<'static, 'static> {
...
+        .subcommand(
+            App::new("run")
+                .about("Run the relay")
+                .after_help(
+                    "This runs the relay in the foreground until it's shut down.  It will bind \
              to the port and network interface configured in the config file.",
-        ))
+                )
+                .arg(
+                    Arg::with_name("secret_key")
```
Comment (jan-auer) [inline src/cliapp.rs:37]: The names here are with underscores, and in all other cases with dashes. Can we unify this?
I personally prefer underscores for the name.

File: relay-config/src/config.rs [hunk 16]
```diff
@@ -643,6 +700,160 @@ impl Config {
...
+                    return Err(ConfigError::new_field_error(
+                        "processing",
+                        val,
+                        Context::new(ConfigErrorKind::InvalidOverride),
+                    ));
+                }
+            }
+        }
+
+        // TODO are we going to support Cluster redis ?
```
Comment (jan-auer) [inline relay-config/src/config.rs]: This is fine. It was explicitly defined that redis cluster is not supported for this type of configuration. Feel free to remove this TODO.

File: relay-config/src/config.rs [hunk 17]
```diff
@@ -643,6 +700,160 @@ impl Config {
...
+        if let Some(redis) = overrides.redis_url {
+            processing.redis = Some(RedisConfig::Single(redis))
+        }
+
+        if let Some(kafka_url) = overrides.kafka_url {
+            let mut bootstrap_servers_found = false;
+            for entry in processing.kafka_config.iter_mut() {
+                if entry.name.as_str() == "bootstrap.servers" {
+                    entry.value = kafka_url.clone();
```
Comment (jan-auer) [inline relay-config/src/config.rs]: To avoid this clone, you could use something like this (also, no need to convert to `str`):

```rust
let existing = processing.kafka_config.iter_mut()
  .find(|e| e.name == "bootstrap.servers");

if let Some(entry) = existing {
  entry.value = kafka_url;
} else {
  processing.kafka_config.push(...);
}
```

File: relay-config/src/config.rs [hunk 18]
```diff
@@ -643,6 +700,160 @@ impl Config {
...
+    /// Override configuration with values coming from other sources (e.g. env variables or
+    /// command line parameters)
+    pub fn override_config(
+        &mut self,
+        overrides: OverridableConfig,
+    ) -> Result<&mut Self, ConfigError> {
+        let relay = &mut self.values.relay;
+
+        if let Some(upstream) = overrides.upstream {
+            relay.upstream = UpstreamDescriptor::from_str(&upstream).map_err(|err| {
```
Comment (jan-auer) [inline relay-config/src/config.rs]: Instead of importing `FromStr`, it would be more idiomatic to call parse here. The type can be inferred by Rust.

```rust
relay.upstream = upstream.parse().map_err(...);
```

**Review by jan-auer — CHANGES_REQUESTED** [review]
I really like the approach and general implementation. This is a good and extensible way to pass overrides around. There are a few stylistic suggestions below.

For a moment, I was wondering whether `OverridableConfig` should already contain the parsed types. In turn, that would mean, that the values need to be parsed outside of `fn override`. This has a pro and a con:
- **Pro**: We could differentiate in errors between an ENV override and an args override. This would maybe make it more clear when both are declared, which is the offending value. 
- **Con**: Parsing would have to be duplicated.

In the end, I think the approach you chose is the better one.

File: relay-config/src/config.rs (hunk 8 above)
Comment (jan-auer) [reply relay-config/src/config.rs:121]: It seems that this is also used for the env var, which is passed directly in. In that case, this of course needs to stay a string.

File: relay-config/src/config.rs (hunk 5 above)
Comment (RaduW) [reply relay-config/src/config.rs]: `override` is a reserved keyword in Rust. changed to apply_override

File: relay-config/src/config.rs (hunk 6 above)
Comment (RaduW) [reply relay-config/src/config.rs]: ok

File: relay-config/src/config.rs (hunk 7 above)
Comment (RaduW) [reply relay-config/src/config.rs]: ok

File: relay-config/src/config.rs (hunk 8 above)
Comment (RaduW) [reply relay-config/src/config.rs:121]: Initially it was a bool, and port was an i16 but I decided finally to make them all Option<String> in order to group all conversions in one place (i.e. in apply_override) instead of doing the conversions (and especially dealing with the errors) in both environment variables and parameters overrides.

File: relay-config/src/config.rs (hunk 10 above)
Comment (RaduW) [reply relay-config/src/config.rs]: ok

File: src/cli.rs (hunk 11 above)
Comment (RaduW) [reply src/cli.rs:101]: I personally don't like fuzzy configuration but as you wish.

File: src/cliapp.rs (hunk 12 above)
Comment (RaduW) [reply src/cliapp.rs]: ok

File: src/cliapp.rs (hunk 13 above)
Comment (RaduW) [reply src/cliapp.rs]: Not that this is not an optional parameter but a flag, there are two: 
` --processing-enabled` and `--processing-disabled` it makes it clear what you are specifying, if you want to save your keyboard from wear you can also use `-e` or 
`-d`. 

I think it is nicer to have flags for boolean values rather than parameters with value 
`--processing true`, or `--processing 1` .

Any way if you still think it is a good idea I can change it to either a parameter 
`--processing true | 1`  or a flag  `--processing` and `--no-processing`. 

Note that using `--processing` will make the short option unintuitive because `p` is already used twice `-p` for `--public-key` and `-P` for `--port`

File: src/cliapp.rs (hunk 14 above)
Comment (RaduW) [reply src/cliapp.rs]: see comment above about my preference. Note that there are already mutually exclusive since they are in an arg group:

                .arg(
                    Arg::with_name("processing-enabled")
                        .short("e")
                        .long("processing-enabled")
                        .help("Enable processing."),
                )
                .arg(
                    Arg::with_name("processing-disabled")
                        .short("d")
                        .long("processing-disabled")
                        .help("Disable processing."),
                )
                .group(
                    ArgGroup::with_name("processing")
                        .args(&["processing-enabled", "processing-disabled"])
                        .multiple(false),
                )

File: src/cliapp.rs (hunk 15 above)
Comment (RaduW) [reply src/cliapp.rs:37]: Ok, I prefer the names to be the same as the parameters, in this case they are with underscore to keep it identical with the code used in the credentials sub command. Also the name is only internal to the code **but you do have a point, it should be consistent**.
I'll change all internal names to use underscores.

File: relay-config/src/config.rs (hunk 18 above)
Comment (RaduW) [reply relay-config/src/config.rs]: cool, changed but I do need to specify the Error type in map_err (not sure exactly why).

File: relay-config/src/config.rs (hunk 17 above)
Comment (RaduW) [reply relay-config/src/config.rs]: nice

File: relay-config/src/config.rs (hunk 9 above)
Comment (RaduW) [reply relay-config/src/config.rs]: not very clear, need to discuss further

Comment (RaduW) [reply relay-config/src/config.rs]: At the end the signature is:
``` rust
    fn new_field_error<E>(name: &'static str, inner: E) -> Self
    where
        E: Fail
```
Making it :
``` rust
    fn new_field_error<E>(name: &'static str, inner: E) -> Self
    where
        E: Display + Send + 'static
```
Would also work but I would loose the stack information for Failures that have stack informations ( like Errors).

File: relay-config/src/config.rs [hunk 19]
```diff
@@ -643,6 +692,121 @@ impl Config {
...
+        let id = if let Some(id) = overrides.id {
+            let id = Uuid::parse_str(&id).map_err(|err| ConfigError::new_field_error("id", err))?;
+            Some(id)
+        } else {
+            None
+        };
+        let public_key = if let Some(public_key) = overrides.public_key {
+            let public_key = public_key
+                .parse()
+                .map_err(|err: KeyParseError| ConfigError::new_field_error("public_key", err))?;
```
Comment (jan-auer) [inline relay-config/src/config.rs]: _nit_: This can be inferred by the compiler. Same below for `secret_key`.

File: relay-config/src/config.rs [hunk 20]
```diff
@@ -643,6 +692,121 @@ impl Config {
...
+        let relay = &mut self.values.relay;
+
+        if let Some(upstream) = overrides.upstream {
+            relay.upstream = upstream
+                .parse()
+                .map_err(|err: UpstreamParseError| ConfigError::new_field_error("upstream", err))?;
+        }
+
+        if let Some(host) = overrides.host {
+            //relay.host = IpAddr::from_str(&host).map_err(|err| {
```
Comment (jan-auer) [inline relay-config/src/config.rs]: Please remove.

File: relay-config/src/config.rs [hunk 21]
```diff
@@ -3,47 +3,63 @@ use std::fmt;
...
 #[derive(Debug)]
 pub struct ConfigError {
-    filename: PathBuf,
+    error_source: ErrorSource,
     inner: Context<ConfigErrorKind>,
 }
 
 impl ConfigError {
-    fn new<P: AsRef<Path>>(p: P, inner: Context<ConfigErrorKind>) -> Self {
+    fn new_file_error<P: AsRef<Path>>(p: P, inner: Context<ConfigErrorKind>) -> Self {
```
Comment (jan-auer) [inline relay-config/src/config.rs]: _nit_: Since the type is named `*Error`, it's uncommon to have `_error` as suffix. Personally, I would go with `for_file` and `for_field`.

File: relay-config/src/config.rs [hunk 22]
```diff
@@ -3,47 +3,63 @@ use std::fmt;
...
             Ok(val) => val,
-            Err(err) => return Err(ConfigError::new($path, err.context($kind))),
+            Err(err) => return Err(ConfigError::new_file_error($path, err.context($kind))),
         }
     };
 }
 
+/// Defines the source of a config error
+#[derive(Debug)]
+enum ErrorSource {
```
Comment (jan-auer) [inline relay-config/src/config.rs:31]: _nit_: The Rust convention for such types is to call them "Kind". I would call this `ConfigErrorKind`.

File: relay-config/src/config.rs [hunk 23]
```diff
@@ -87,6 +108,34 @@ pub enum ConfigErrorKind {
     /// compiled without the processing feature.
     #[fail(display = "was not compiled with processing, cannot enable processing")]
     ProcessingNotAvailable,
+    /// An override from a environment variable or a command line arg could not be
+    /// parsed into the proper type
+    #[fail(display = "Invalid override")]
```
Comment (jan-auer) [inline relay-config/src/config.rs]: _nit_: Please lowercase for consistency with other messages. 

Can we make this message more descriptive for users who have no concept of "overrides" and are just providing a config value? We could even use the `InvalidValue` variant or the same message, since it is really the same to users. The inner error will then have more information.

File: relay-config/src/config.rs [hunk 24]
```diff
@@ -643,6 +692,121 @@ impl Config {
...
+                (Some(id), Some(public_key), Some(secret_key)) => {
+                    self.credentials = Some(Credentials {
+                        id,
+                        public_key,
+                        secret_key,
+                    })
+                }
+                (None, None, None) => {
+                    // nothing provided, we'll just leave the credentials None, maybe we
+                    // don't need them in the current command or we'll override them latter
```
Comment (jan-auer) [inline relay-config/src/config.rs]:
```suggestion
                    // don't need them in the current command or we'll override them later
```

File: relay-config/src/config.rs [hunk 25]
```diff
@@ -643,6 +692,121 @@ impl Config {
...
+        if let Some(port) = overrides.port {
+            relay.port = u16::from_str_radix(port.as_str(), 10)
+                .map_err(|err| ConfigError::new_field_error("port", err))?;
+        }
+
+        let processing = &mut self.values.processing;
+        if let Some(enabled) = overrides.processing {
+            match enabled.to_lowercase().as_str() {
+                "true" | "1" => processing.enabled = true,
+                _ => processing.enabled = false,
```
Comment (jan-auer) [inline relay-config/src/config.rs]: I see your point that you made in [this comment](https://github.com/getsentry/relay/pull/546/files#r409474592). Would you prefer to only set to false if there is `"false" | "0"` given explicitly?

File: relay-config/src/config.rs (hunk 9 above)
Comment (jan-auer) [reply relay-config/src/config.rs]: That makes sense, thanks! Let's keep `Fail` here for now and reevaluate, once we're moving to `Error`.

File: relay-config/src/config.rs (hunk 19 above)
Comment (RaduW) [reply relay-config/src/config.rs]: ok

File: relay-config/src/config.rs (hunk 20 above)
Comment (RaduW) [reply relay-config/src/config.rs]: yep

File: relay-config/src/config.rs (hunk 21 above)
Comment (RaduW) [reply relay-config/src/config.rs]: ok

File: relay-config/src/config.rs (hunk 22 above)
Comment (RaduW) [reply relay-config/src/config.rs:31]: We already have a `ConfigErrorKind` with values like CouldNotOpenFile, BadYaml and InvalidOverride ... for all except InvalidOverride the source is a configuration file, for InvalidOverride the source is an Environment Variable or a CLI parameter. The Error source is defined to keep details specific to the source of error (i.e. file or env var).

File: relay-config/src/config.rs (hunk 24 above)
Comment (RaduW) [reply relay-config/src/config.rs]: ok

File: relay-config/src/config.rs (hunk 23 above)
Comment (RaduW) [reply relay-config/src/config.rs]: Ok, removed and replaced usages with InvalidValue ( I still think that there is some value in having a dedicated ErrorKind for invalid overrides).

File: relay-config/src/config.rs (hunk 25 above)
Comment (RaduW) [reply relay-config/src/config.rs]: Yes, I would prefer that, I set `"true"| "1"` to `true`  and `"false" | "0" | ""` to `false` , anything else generates an error.

File: relay-config/src/config.rs (hunk 22 above)
Comment (jan-auer) [reply relay-config/src/config.rs:31]: Gotcha. `source` may be a bit misleading because of [`Error::source`](https://doc.rust-lang.org/stable/std/error/trait.Error.html#method.source). Let's leave it for now, though 👍

**Review by jan-auer — APPROVED** [review]
Great patch, thanks!
